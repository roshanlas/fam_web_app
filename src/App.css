body {
  font-family: 'Open Sans', sans-serif !important;
  color: #263633 !important;
  background-color: #F1E4E1;
}
input[type="button"], button, a {
  font-family: 'Oswald', sans-serif !important;
}
.App {
  text-align: center;
}

.App-logo {
  animation: App-logo-spin infinite 20s linear;
  height: 40vmin;
  pointer-events: none;
}

.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

.MuiBottomNavigationAction-root.Mui-selected {
  color: #630663 !important;
}


h1, h2, h3, h4 {
  font-family: 'Oswald', sans-serif !important;
  font-weight: normal;
}
h1 {
  text-align: center;
  color: #707070 !important;
}

h2 {
  margin-top: 0px;
}

h3, h4 {
  margin: 0px;
}