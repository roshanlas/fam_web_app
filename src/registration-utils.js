export const fields = {
    email: {comp: null, error: false, label: 'Email', required: true}, 
    password: {comp: null, error: false, label: 'Password', required: true}, 
    repeatPassword: {comp: null, error: false, label: 'Repeat Password', required: true}, 
    firstName: {comp: null, error: false, label: 'First Name', required: true}, 
    lastName: {comp: null, error: false, label: 'Last Name', required: true},
    dob: {comp: null, error: false, label: 'Date of Birth dd/mm/yyyy', required: true}, 
    gender: {comp: null, error: false, label: 'Gender', required: true}, 
    marriageStatus: {comp: null, error: false, label: 'Marriage Status', required: false}, 
    occupation: {comp: null, error: false, label: 'Occupation', required: false}, 
    residence: {comp: null, error: false, label: 'Residence', required: false},
    country: {comp: null, error: false, label: 'Country', required: true}, 
    city: {comp: null, error: false, label: 'City', required: true},
    homeAddress: {comp: null, error: false, label: 'Home Address', required: false}, 
    postCode: {comp: null, error: false, label: 'Post Code', required: false}
  };

  
/**
 * Converts an object to another type
 * @param {*} formData
 * @param {*} fields
 */
export const populateFormData = (fields) => {
    let formData = {};
    Object.keys(fields).forEach(
        f=>Object.assign(formData, { [`${f}`]: fields[`${f}`].comp.value })
    )
    return formData;
}
  
/**
 * 
 * @param {*} formData 
 */
export const validateEmail = (email) => {
    var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    return re.test(email);
}
  
export const validateFields = (formData) => {
    let errors = {};
    if(!validateEmail(formData['email'])) {
        Object.assign(errors, {'email':'Please enter a valid email address'})
    }
    if(formData['password'] !== formData['repeatPassword']) {
        Object.assign(errors, {'password':'The passwords do not match'})
        Object.assign(errors, {'repeatPassword':'The passwords do not match'})
    }
    if(formData['password'].length === 0 || formData['repeatPassword'].length === 0) {
        Object.assign(errors, {'password':'Please enter a valid password'})
        Object.assign(errors, {'repeatPassword':'Please enter a valid password'})
    }
    if(formData['firstName'].length === 0) {
        Object.assign(errors, {'firstName':'Please enter your first name'})
    }
    if(formData['lastName'].length === 0) {
        Object.assign(errors, {'lastName':'Please enter your last name'})
    }
    if(formData['dob'].length === 0) {
        Object.assign(errors, {'dob':'Please enter your date of birth'})
    }
    if(formData['city'].length === 0) {
        Object.assign(errors, {'city':'Please enter your city'})
    }
    if(formData['country'].length === 0) {
        Object.assign(errors, {'country':'Please select your country'})
    }
    return errors;
}

/**
 * Performs fetch request and invokes callback
 * @param {*} formData
 */
export const submitData = async (formData, service) => {
    return await fetch(
        // URL
        `${process.env.REACT_APP_API_URL}/${service}`, 
        // Data
        {
            method: 'POST',
            body: JSON.stringify(formData),
            headers: {
                "Content-Type": "application/json",
                "Authorization": `Bearer ${localStorage.token}`
            }
        }
    )
}

export const getData = async (service) => {
    return await fetch(
        // URL
        `${process.env.REACT_APP_API_URL}/${service}`,
        {
            headers: {
            "Content-Type": "application/json",
            "Authorization": `Bearer ${localStorage.token}`
            }
        }
    )
}

export default validateEmail;