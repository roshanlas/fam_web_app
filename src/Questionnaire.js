import React, { useState, useContext, useEffect } from 'react';
import CssBaseline from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import Logo from './Logo';
import { BackButton } from './Heading';
import { Link as RouterLink } from 'react-router-dom';
import { submitData, getData } from './registration-utils';
import Button from '@material-ui/core/Button';
import { makeStyles } from '@material-ui/core/styles';
import DoneIcon from '@material-ui/icons/Done';
import colors from './colorTheme';
import { AppContext } from './App';

const useStyles = makeStyles({
    questionnaire: {
        background: colors.pink2,
        backgroundSize: 'cover',
        backgroundPosition: 'center',
        height: '100%',
        minHeight: '100vh',
        color: 'white'
    },
    backButton: {
        top: '1.5rem',
        left: '1.5rem',
        height: '1.6rem',
        width: '1.6rem',
        display: 'block',
        position: 'absolute'
    },
    infoSubHead: {
        display: 'flex',
        justifyContent: 'space-between',
        width: '100%',
        margin: '1rem 0'
    },
    heading: {
        marginBottom: '1.6rem'
    },
    textbox: {
        padding: '1rem',
        fontSize: '1.2rem',
        borderRadius: '0.6rem',
        width: '100%',
        minHeight: '7.5rem',
        '&:focus': {
            outline: 'none'
        }
    },
    button: {
        color: 'white',
        width: '9rem',
        display: 'block',
        margin: '0.6rem 0 0 0.6rem',
        padding: '0.4rem 0rem',
        fontSize: '1rem',
        maxWidth: '12rem',
        fontFamily: 'Oswald, sans-serif !important',
        borderRadius: '.8rem',
        textTransform: 'none',
        backgroundColor: colors.g3,
        '&:hover': {
          backgroundColor: colors.g3,
        }
    },
    loadingButton: {
        color: 'white',
        width: '9rem',
        display: 'block',
        margin: '0.6rem 0 0 0.6rem',
        padding: '0.4rem 0rem',
        fontSize: '1rem',
        maxWidth: '12rem',
        fontFamily: 'Oswald, sans-serif !important',
        borderRadius: '.8rem',
        textTransform: 'none',
        textAlign: 'center',
        backgroundColor: colors.g3,
        opacity: '0.5',
        '&:hover': {
          backgroundColor: colors.g3,
        }
    },
    doneIcon: {
        position: 'absolute',
        top: '0.5rem',
        right: '0.9rem'
    }, 
    error:{
        fontSize: '1.2em',
        color: 'white',
        background: 'red',
        padding: '0.2em 0.6em',
        borderRadius: '0.2em'
    },
    prompt: {
        display: 'none'
    },
    open: {
        display: 'block',
        position: 'fixed',
        top: '0px',
        left: '0px',
        width: '100%',
        height: '100%',
        background: 'rgba(0,0,0,0.2)'
    },
    promptBox: {
        maxWidth: '30em',
        width: 'calc(100% - 2em)',
        padding: '2em',
        position: 'absolute',
        color: 'black',
        background: 'white',
        top: '30%',
        left: '50%',
        transform: 'translate(-50%,-50%)',
        boxShadow: '1px 1px 10px rgba(0,0,0,0.5)'
    }
});

const Questionnaire = (props) => {
    const classes = useStyles();
    const [ globalState, setGlobalState ] = useContext(AppContext);
    const [ state, setState ] = useState({ 
        draftSaved: false, 
        loading: false,
        error: false,
        prompt: 'hidden',
        submit: '',
        answersLoading: true
    });
    let questions = globalState.questions? globalState.questions : [];
    let count =0;
    let fields = [];
    let answers;

    const getStudentSubmission = () => {
        getData(
            `submission/retrieve/${globalState.storyID}`, 
            'GET'
        )
        .then(async res => {
            let ret = await res.json();
            if(res.ok) {
                setGlobalState({
                    ...globalState,
                    submission: ret.submission === null ? [] : ret.submission
                });
            } else {
                setGlobalState({
                    ...globalState,
                    submission: false
                })
            }
        })
        .catch(err => {
            console.log('err', err)
            setGlobalState({
                ...globalState,
                submission: false
            })
        });
    };

    /**
     * saveDraft
     */
    const saveDraft = () => {

        setState({ 
            ...state, 
            draftSaved: true, 
            loading: true
        });

        answers = fields.map(field=>
            { return {answer: field.value} }
        );

        submitData({
            submission: answers, 
            storyID: globalState.storyID,
            final: false
        }, 'submission/submit')
        .then(async res=>{
            let ret = Promise.resolve(res.json());
            if(res.ok) {
                setState(
                    { ...state, draftSaved: true, loading: false }
                );
            } else {
                setState(
                    { 
                        ...state, 
                        draftSaved: false, 
                        loading: false, 
                        error: true
                    }
                );
            }
        }).catch(err => {
            setState(
                { 
                    ...state, 
                    draftSaved: false, 
                    loading: false,
                    error: true
                }
            );
        })
    };

    /**
     * 
     */
    const promptSave = () => {
        setState({
            ...state,
            prompt: 'open'
        });
    };

    /**
     * 
     */
    const closePrompt = () => {
        setState({
            ...state,
            prompt: 'close'
        });
    }

    /**
     * submitAndConfirm
     */
    const submitAndConfirm = () => {

        setState({ 
            ...state, 
            draftSaved: true, 
            loading: true
        });

        answers = fields.map(field=>
            { return {answer: field.value} }
        );

        submitData({
            submission: answers, 
            final: true,
            storyID: globalState.storyID,
        }, 'submission/submit')
        .then(async res=>{
            let ret = Promise.resolve(res.json());
            if(res.ok) {
                setState({ 
                    ...state, 
                    loading: false,
                    prompt: 'hidden',
                    submit: 'done'
                });
            } else {
                setState({ 
                    ...state, 
                    loading: false, 
                    error: true,
                    prompt: 'hidden',
                    submit: ''
                });
            }
        }).catch(err => {
            setState(
                { 
                    ...state, 
                    loading: false,
                    error: true,
                    prompt: 'hidden',
                    submit: ''
                }
            );
        })
    };


    useEffect(()=>{
        
        if(globalState.submission === undefined) {
            getStudentSubmission();
        }
    
        if(questions.length === 0) {
            props.history.push(`/profile`)
        }
        count = 0;
    });


    return(
        <div className={`${classes.questionnaire}`}>
            <CssBaseline />
            <div className={classes.header}>
                <BackButton to="/profile" />
                <Logo />
            </div>
            <Container maxWidth="xs">
                <div className={classes.infoSubHead}>
                    <h3>Day {globalState.dayOfMonth} Story of the Day</h3>
                    <h3>{globalState.occupation}</h3>
                </div>
                <h2 className={classes.heading}>{globalState.person}</h2>
                
                {globalState.submission && questions.map(question=>{
                        var e = <div key={count}>
                                <h2>{`Question ${count+1}:`}</h2>
                                <p>{question.title}</p>
                                <textarea 
                                ref={comp=>fields.push(comp)}
                                className={classes.textbox} 
                                defaultValue={
                                    globalState.submission.length > 0 ? 
                                        globalState.submission[count].answer : ''
                                }/>
                            </div>
                        count++; return e;
                    }
                )} 

                
                <div style={{ display: 'flex', justifyContent: 'flex-end', marginBottom: '2em' }}>
                    { !state.loading && 
                        <Button 
                            onClick={saveDraft}
                            st={state.draftSaved.toString()}
                            className={`btn-draft ${classes.button}`}>
                                Save Draft {state.draftSaved && <DoneIcon className={classes.doneIcon}/>}
                        </Button> 
                    }
                    { state.loading && <div className={`${classes.loadingButton}`}>Please wait</div>}
                    <Button 
                        onClick={promptSave}
                        className={`btn-submit ${classes.button} ${state.submit}`}>
                            Submit
                    </Button>
                </div>
                { state.error && <p className={classes.error}>An error occured. Please try again.</p>}

                <div className={`${classes.prompt} ${classes[state.prompt]} prompt ${state.prompt}`}>
                    <div className={classes.promptBox}>
                        <p>Are you sure you want to submit your answers?</p>
                        <p>You will not be able to change your answers later.</p>
                        <Button className="btn-confirm" onClick={submitAndConfirm}>Submit</Button>
                        <Button className="btn-cancel" onClick={closePrompt}>Cancel</Button>
                    </div>
                </div>
            </Container>
        </div>
    )
}

export default Questionnaire;