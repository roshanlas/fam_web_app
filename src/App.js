import React, { useState, createContext, useEffect } from 'react';
import {  
    BrowserRouter as Router,
    Route,
    Switch
} from 'react-router-dom';
import { PrivateRoute } from './PrivateRoute';
import Main from './Main';
import SignUp from './SignUp';
import SignIn from './SignIn';
import About from './About';
import Challenge from './Challenge';
import TermsConditions from './TermsConditions';
import Profile from './Profile';
import ForgotPassword from './ForgotPassword';
import Questionnaire from './Questionnaire';
import Story from './Story';
import './App.css';

export const AppContext = createContext();

const App = () => {

  const [state, setState] = useState({
    user: {
      firstName: null,
      lastName: null,
      token: null,
    },
    loginStatus: localStorage.token ? true : false,
  });

  useEffect(()=>{
    switch(state.loginStatus) {
      case true:
          // Put the JWT in the browser storage
          localStorage.setItem('firstName', state.firstName);
          localStorage.setItem('lastName', state.lastName);
          localStorage.setItem('token', state.token);
        break;
      case false:
          localStorage.clear();
        break;
      default:
        break;
    }   
  })

  return (
  <AppContext.Provider value={[state, setState]} >    
    <Router>
        <Switch>
            <Route path="/" exact component={Main}/>
            <Route path="/signup" component={SignUp} />
            <Route path="/sign-in" component={SignIn} />
            <Route path="/forgot-password" component={ForgotPassword} />
            <Route path="/about" component={About} />
            <Route path="/30-day-challenge" component={Challenge} />
            <Route path="/terms-and-conditions" component={TermsConditions} />
            <PrivateRoute path="/profile" component={Profile} />
            <PrivateRoute path="/questionnaire" component={Questionnaire} />
            <PrivateRoute path="/story" component={Story} />
        </Switch>
    </Router>
  </AppContext.Provider>    
  );
}

export default App;
