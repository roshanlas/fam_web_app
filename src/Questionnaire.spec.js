import React from 'react';
import {  
    MemoryRouter,
} from 'react-router';
import { configure, shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { submitData, getData } from './registration-utils';
import { AppContext } from './App'
import Questionnaire from './Questionnaire';
import { act } from 'react-dom/test-utils';

jest.mock('./registration-utils', () => ({
    submitData: jest.fn().mockImplementation(
        () => Promise.resolve(
            {
                ok: true, 
                json: jest.fn().mockResolvedValue(true)
            }
        )
    ),
    getData: jest.fn().mockImplementation(
        () => Promise.resolve(
            {
                ok: true, 
                json: jest.fn().mockResolvedValue(true)
            }
        )
    )
}))

configure({ adapter: new Adapter() });

describe('render', ()=> {
    let wrapper0;
    let wrapper;

    // const setState = jest.fn();
    // const useStateSpy = jest.spyOn(registrationUtils, 'submitData')
    // useStateSpy.mockImplementation((init) => {ok: true} );
    const props = {
        history: {
            push: jest.fn()
        }
    }


    beforeEach(()=>{
        wrapper0 = mount(
            <AppContext.Provider value={[{}, ()=>{}]}>
                <MemoryRouter>
                    <Questionnaire {...props}/>
                </MemoryRouter>
            </AppContext.Provider>
        );
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it('should render without crashing',()=>{
        expect(wrapper).toMatchSnapshot();
    });

    it('should set saved to true if draft button clicked', async ()=> {  
        await act(async () => {
            await wrapper0.find('.btn-draft').at(0).simulate('click', {target: { value: 'Foo' } })
            expect(wrapper0.update().find('.btn-draft').at(0).props().st).toBe("true");
        });
    });

    it('should prompt user when submit is clicked', async () => {
        await act(async () => {
            await wrapper0.find('.btn-submit').at(0).simulate('click');
            expect(wrapper0.update().find('.prompt.open').at(0)).toHaveLength(1);
        });
    });

    it('should close prompt when cancel is clicked', async () => {
        await act(async () => {
            await wrapper0.find('.btn-submit').at(0).simulate('click');
            await wrapper0.find('.btn-confirm').at(0).simulate('click');
            expect(wrapper0.update().find('.btn-submit.done').at(0)).toHaveLength(1);
        });
    });

    it('should save when confirm is clicked', async () => {
        await act(async () => {
            await wrapper0.find('.btn-submit').at(0).simulate('click');
            await wrapper0.find('.btn-confirm').at(0).simulate('click');
            expect(wrapper0.update().find('.prompt.open').at(0)).toHaveLength(0);
        });
    });

});