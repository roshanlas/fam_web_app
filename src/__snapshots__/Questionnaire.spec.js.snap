// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render should render without crashing 1`] = `undefined`;
